lib/components/Auth/AuthService.ts

```diff
@@ -0,0 +1,139 @@
+import crypto from 'crypto';
+
+import TypeService from '../../../src/services/TypeService';
+import MOST_COMMON_PASSWORDS from './MOST_COMMON_PASSWORDS';
+import UtilityService from '../../../src/services/UtilityService';
+
+const MIN_PASSWORD_LENGTH = 8;
+const MAX_PASSWORD_LENGTH = 130;
+
+const CONFIG = {
+    ENCRYPTION: {
+        ITERATIONS: 200 * 1000,
+        SALT_LENGTH: 32,
+        METHOD: 'sha256',
+        OUTPUT_LENGTH: 32
+    }
+};
+
+//--==--==----==--==--==--==----==--==----==--==----==--==--==--==----==--==--//
+
+
+export function validatePassword(password): boolean | string {
+    if (TypeService.isString(password, true) === false) {
+        return 'password (String) is required and cannot be empty';
+    } else if (password.length < MIN_PASSWORD_LENGTH) {
+        return 'password must be at least ' + MIN_PASSWORD_LENGTH + ' characters long';
+    } else if (password.length > MAX_PASSWORD_LENGTH) {
+        return 'password cannot be more than ' + MAX_PASSWORD_LENGTH + ' characters';
+    } else if (passwordIsInListOfMostCommonPasswords(password)) {
+        return 'password is in our list of 10,000 most commonly used passwords, please try again';
+    } else {
+        return true;
+    }
+}
+
+
+
+export function encryptPassword(password: string, iterations?: number, salt?: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+        if (TypeService.isString(password, true) === false) {
+            reject(new Error('password (String) is required'));
+        } else {
+            iterations = (iterations) ? iterations : CONFIG.ENCRYPTION.ITERATIONS;
+            salt = (salt) ? salt : generateSalt();
+
+            const hashMethod = CONFIG.ENCRYPTION.METHOD;
+            const outputLength = CONFIG.ENCRYPTION.OUTPUT_LENGTH;
+
+            crypto.pbkdf2(password, salt, iterations, outputLength, hashMethod, (error, derivedKey) => {
+                if (error) {
+                    reject(error);
+                } else {
+                    const storageValue = generateStorablePasswordString(iterations, salt, derivedKey);
+                    resolve(storageValue);
+                }
+            });
+        }
+    });
+}
+
+
+
+export function comparePassword(unencryptedPassword: string, encryptedPassword: string): Promise<boolean> {
```

Comment:
This compares the user-input password and the encrypted password. It resolves the comparison, which could be true or false, it doesn't reject if the passwords do not match

Reply:
This is the flipside to the promise resolve/reject question. This one resolves a boolean, only when the function has run as expected, and rejects only when an error occurs, not when the passwords don't match